src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt

```diff
@@ -39,19 +36,13 @@ public class DependencyManager @Inject constructor(val executors: KobaltExecutor
      * Reorder dependencies so that if an artifact appears several times, only the one with the higest version
      * is included.
      */
+    // TODO it may cause problems if there are file dependencies in the list, we can't compare file dependencies with maven dependencies
     public fun reorderDependencies(dependencies: Collection<IClasspathDependency>): List<IClasspathDependency> {
-        val result = arrayListOf<IClasspathDependency>()
-        val map : ArrayListMultimap<String, IClasspathDependency> = ArrayListMultimap.create()
-        // The multilist maps each artifact to a list of all the versions found
-        // (e.g. {org.testng:testng -> (6.9.5, 6.9.4, 6.1.1)}), then we return just the first one
-        dependencies.forEach {
-            map.put(it.shortId, it)
-        }
-        for (k in map.keySet()) {
-            val l = map.get(k)
-            Collections.sort(l, Collections.reverseOrder())
-            result.add(l.get(0))
-        }
-        return result
+        return dependencies.groupBy { it.shortId + ":" + it.classifierIfAvailable() }
```

Comment:
Then maybe let's remove this from the PR so you can look into it separately
if you want?

Overall, I think this PR could have been broken in smaller different PR's,
which would make merging them individually much easier. Just a thought for
your future PR's :)



-- 
Cédric


On Fri, Nov 13, 2015 at 6:00 AM, Sergey Mashkov <notifications@github.com>
wrote:

> In src/main/kotlin/com/beust/kobalt/maven/DependencyManager.kt
> <https://github.com/cbeust/kobalt/pull/38#discussion_r44784167>:
>
> >      public fun reorderDependencies(dependencies: Collection<IClasspathDependency>): List<IClasspathDependency> {
> > -        val result = arrayListOf<IClasspathDependency>()
> > -        val map : ArrayListMultimap<String, IClasspathDependency> = ArrayListMultimap.create()
> > -        // The multilist maps each artifact to a list of all the versions found
> > -        // (e.g. {org.testng:testng -> (6.9.5, 6.9.4, 6.1.1)}), then we return just the first one
> > -        dependencies.forEach {
> > -            map.put(it.shortId, it)
> > -        }
> > -        for (k in map.keySet()) {
> > -            val l = map.get(k)
> > -            Collections.sort(l, Collections.reverseOrder())
> > -            result.add(l.get(0))
> > -        }
> > -        return result
> > +        return dependencies.groupBy { it.shortId + ":" + it.classifierIfAvailable() }
>
> No, it is not equivalent. First of all notice that Collections.sort(l,
> Collections.reverseOrder()) in fact is not correct. Perhaps it should be
> compilation error at this place. l is a list of IClasspathDependency that
> is not comparable and it is not clear what is going to happen if there are
> both maven and file dependencies.
>
> Also now grouping key consist of shortId and classifier (if exist).
> In fact it need to be tested but both maven and file dependency classes
> requires too much infrastrucuture. However I will try
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cbeust/kobalt/pull/38/files#r44784167>.
>